# DeSeq2 Module Tutorial
## Differential Gene Expression Analysis Using **[NeatSeq-Flow](http://neatseq-flow.readthedocs.io/en/latest/)** Platform
&nbsp;  
&nbsp;
&nbsp;  
<img align="right" src="https://raw.githubusercontent.com/bioinfo-core-BGU/NeatSeq-Flow_Workflows/master/DeSeq_Workflow/doc/DeSeq2_Module_Report.gif" width="1050">
&nbsp;  
&nbsp;
&nbsp;  
&nbsp;
&nbsp;  


***
## This Module Can Perform:
 - Differential Gene Expression Using 'DESeq2' R Package
 - Gene Annotation from Biomart or from Trinotate Results for De-Novo Transcriptome Assembly
 - Quality control eg. MA,Volcano and PCAs
 - Genes And Samples Filtering using 'scater' R package
 - Expression Patterns Clustering of Significant Genes
 - Clusters Visualization eg. Heatmaps and Trend Plots
 - Gene Ontology and KEGG Enrichment Analysis
 - Enrichment Analysis Visualization eg. Dot Plots
 - Gene Ontology and KEGG Terms Genes Overlap Visualization
 - Generate Final Report in HTML Format


## In this Tutorial we will:
### Use the Public RNA-Seq Data [GSE72232](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE72232), Generated by [Inukai et al. 2018](https://www.ncbi.nlm.nih.gov/pubmed/29114017) 
 - [Download the GSE72232 Data](#download-the-gse72232-data)   
 - [ Installation (Each Step Can be skipped if already installed)](#Installation)
     - [Install Conda](#install-conda)
     - [Install NeatSeq-Flow](#install-neatseq-flow)
     - [Install DeSeq2 conda environment](#Install-DeSeq2-conda-environment)
     - [Install NeatSeq-Flow-GUI](#install-neatseq-flow-gui)
 - [How To Use The DESeq2 Module](#how-to-use-the-deseq2-module)
 - [Start The GUI](#start-the-gui)
    - [Add DeSeq2 Step](#add-deseq2-step)
 




***

## Download the GSE72232 Data

  1. Create New Directory for the Tutorial
       ```Bash
         mkdir GSE72232
         cd GSE72232
       ```
  2. Download The Raw Count Data:
        ```Bash
          wget ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE72nnn/GSE72232/suppl/GSE72232_RNA-seq_combined_counts.txt.gz
        ```
  3. Download The Samples Meta-Data File:
        ```Bash
          wget https://raw.githubusercontent.com/bioinfo-core-BGU/NeatSeq-Flow_Workflows/master/DeSeq_Workflow/Samples_MetaData
        ```
     ### The Samples Meta-Data File:

     | Samples          | Genotype | Age  | Rep  |
     |------------------|:--------:|------|------|
     | N2_Day0_rep1     |    N2    | Day0 | rep1 |
     | N2_Day5_rep1     |    N2    | Day5 | rep1 |
     | mir-71_Day0_rep1 |  mir-71  | Day0 | rep1 |
     | mir-71_Day5_rep1 |  mir-71  | Day5 | rep1 |
     | N2_Day0_rep2     |    N2    | Day0 | rep2 |
     | N2_Day5_rep2     |    N2    | Day5 | rep2 |
     | mir-71_Day0_rep2 |  mir-71  | Day0 | rep2 |
     | mir-71_Day5_rep2 |  mir-71  | Day5 | rep2 |
     | N2_Day0_rep3     |    N2    | Day0 | rep3 |
     | N2_Day5_rep3     |    N2    | Day5 | rep3 |
     | mir-71_Day0_rep3 |  mir-71  | Day0 | rep3 |
     | mir-71_Day5_rep3 |  mir-71  | Day5 | rep3 |

## Install Conda

  - For Linux 64bit, in the terminal:
    ```Bash
      wget https://repo.continuum.io/miniconda/Miniconda2-latest-Linux-x86_64.sh
      sh Miniconda2-latest-Linux-x86_64.sh
    ```
    **During conda's installation: type yes to add conda to the PATH**
    
    For different operating system go to [Conda](https://conda.io/miniconda.html) 
    
## Install NeatSeq-Flow 

  1. Download the **NeatSeq Flow** installer file:
        ```Bash
          wget http://neatseq-flow.readthedocs.io/en/latest/extra/NeatSeq_Flow_conda_env.yaml
        ```
  2. Create the **NeatSeq_Flow** conda environment:
        ```Bash
           conda env create -f NeatSeq_Flow_conda_env.yaml
        ```  
        
## Install DeSeq2 conda environment
  1. Download the **DeSeq2 conda environment** installer file:
        ```Bash
          wget https://raw.githubusercontent.com/bioinfo-core-BGU/NeatSeq-Flow_Workflows/master/DeSeq_Workflow/DeSeq2_module/DeSeq2_env_install.yaml
        ```
  2. Create the **DeSeq2** conda environment:
        ```Bash
          conda env create -f DeSeq2_env_install.yaml
        ```  
   **Note That The Installation Can Take Some Time!!!**
  
        
        
## Install NeatSeq-Flow-GUI
  1. Download the **NeatSeq-Flow-GUI** installer file:
        ```Bash
          wget https://raw.githubusercontent.com/bioinfo-core-BGU/NeatSeq-Flow-GUI/master/NeatSeq_Flow_GUI_installer.yaml
        ```
  2. Create the **NeatSeq_Flow_GUI** conda environment:
        ```Bash
          conda env create -f NeatSeq_Flow_GUI_installer.yaml
        ```  


## How To Use The DESeq2 Module 

    
## Start The GUI
  1. Activate the **NeatSeq_Flow_GUI** conda environment:
        ```Bash
          bash
          source activate NeatSeq_Flow_GUI
        ```
  2. Run **NeatSeq_Flow_GUI**:
        ```Bash 
          NeatSeq_Flow_GUI.py
        ```


    ### Add DeSeq2 Step
       <img src="https://raw.githubusercontent.com/bioinfo-core-BGU/NeatSeq-Flow_Workflows/master/DeSeq_Workflow/doc/Add_DeSeq2_Step.gif" width="650">
       
       1. **Add New Step:**   
          In the 'Work-Flow' Tab choose The module template "DeSeq2_Simple(Using_Conda)" and click on the 'Create New Step' button.
       2. **Choose Step base:**
          - Click on the "DeSeq2" step button to see the step options. 
          - Click on the "base" option in the left panel of the GUI, just below the step "DeSeq2" name. 
          - Click on the "Value options:" drop-down menu in the upper left area of the GUI. 
          - Choose the "Merge" step to be the previous step for the DeSeq2 new step.
          - Click on the "Add" button.
          - Click on the "Edit" button.
          
       3. **Edit the "redirects" options of the "DESeq2" step**
       
            - **Below the "redirects" in the left panel of the GUI:**
              - Click on the "--DESIGN" option and edit the "Value" field to : '~ Genotype + Age + Genotype:Age' then click the "Edit" button.
              - Click on the "--LRT" option and edit the "Value" field to : '~ Genotype + Age' then click the "Edit" button.
              - Click on the "--CONTRAST" option and edit the "Value" field to : 'Genotype,N2,mir-71|Age,Day0,Day5' then click the "Edit" button.
              - Click on the "--X_AXIS" option and edit the "Value" field to : 'Age' then click the "Edit" button.
              - Click on the "--GROUP" option and edit the "Value" field to : 'Genotype' then click the "Edit" button.
              - Click on the "--PCA_COLOR" option and edit the "Value" field to : 'Genotype' then click the "Edit" button.
              - Click on the "--PCA_SHAPE" option and edit the "Value" field to : 'Age' then click the "Edit" button.
              
            **All values in this section refer to the fields found in [The Samples Meta-Data File](#the-samples-meta-data-file)**

            **For more information about the values possibilities of "--DESIGN","--LRT" and "--CONTRAST" see the [DESeq2 manual](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)**

            **It is possible to indicate more then one contrast by using "|"**
       
# Contact
Please contact Liron Levin at: [levinl@post.bgu.ac.il](mailto:levinl@post.bgu.ac.il)
