# Differential Expression (DESeq2), Clustering and Functional Analyses Tutorial
## Differential Gene Expression Analysis Using **[NeatSeq-Flow](http://neatseq-flow.readthedocs.io/en/latest/)** Platform
&nbsp;  
&nbsp;
&nbsp;  
<img align="right" src="https://raw.githubusercontent.com/bioinfo-core-BGU/NeatSeq-Flow_Workflows/master/DeSeq_Workflow/doc/DeSeq2_Module_Report.gif" width="1050">
&nbsp;  
&nbsp;
&nbsp;  
&nbsp;
&nbsp;  


***
## This Module Can Perform:
 - Differential Gene Expression Using 'DESeq2' R Package
 - Gene Annotation from Biomart or from Trinotate Results for De-Novo Transcriptome Assembly
 - Quality control eg. MA,Volcano and PCAs
 - Genes And Samples Filtering using 'scater' R package
 - Expression Patterns Clustering of Significant Genes
 - Clusters Visualization eg. Heatmaps and Trend Plots
 - Gene Ontology and KEGG Enrichment Analysis
 - Enrichment Analysis Visualization eg. Dot Plots
 - Gene Ontology and KEGG Terms Genes Overlap Visualization
 - Generate Final Report in HTML Format

***
## In this Tutorial we will:
### Use the Public RNA-Seq Data [GSE72232](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE72232), Generated by [Inukai et al. 2018](https://www.ncbi.nlm.nih.gov/pubmed/29114017) 
 - [Download the GSE72232 Data](#download-the-gse72232-data)   
 - [ Installation (Each Step Can be skipped if already installed)](#Installation)
     - [Install Conda](#install-conda)
     - [Install NeatSeq-Flow](#install-neatseq-flow)
     - [Install DeSeq2 conda environment](#Install-DeSeq2-conda-environment)
     - [Install NeatSeq-Flow-GUI](#install-neatseq-flow-gui)
 - [How To Use The DESeq2 Module](#how-to-use-the-deseq2-module)
 - [Start The GUI](#start-the-gui)
    - [Add DeSeq2 Step](#add-deseq2-step)
    - [Possible Options of The "DESeq2" Mudule](#possible-options-of-the-deseq2-mudule)
    - [Edit The Vars Tab](#edit-the-vars-tab)
    - [Create a Samples File](#create-a-samples-file)
    - [Run Work-Flow](#run-work-flow)
 - [Contact](#contact)

***

## Download the GSE72232 Data

  1. Create New Directory for the Tutorial
       ```Bash
         mkdir GSE72232
         cd GSE72232
       ```
  2. Download The Raw Count Data:
        ```Bash
          wget ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE72nnn/GSE72232/suppl/GSE72232_RNA-seq_combined_counts.txt.gz
        ```
  3. Download The Samples Meta-Data File:
        ```Bash
          wget https://raw.githubusercontent.com/bioinfo-core-BGU/NeatSeq-Flow_Workflows/master/DeSeq_Workflow/Samples_MetaData
        ```
     ### The Samples Meta-Data File:

     | Samples          | Genotype | Age  | Rep  |
     |------------------|:--------:|------|------|
     | N2_Day0_rep1     |    N2    | Day0 | rep1 |
     | N2_Day5_rep1     |    N2    | Day5 | rep1 |
     | mir-71_Day0_rep1 |  mir-71  | Day0 | rep1 |
     | mir-71_Day5_rep1 |  mir-71  | Day5 | rep1 |
     | N2_Day0_rep2     |    N2    | Day0 | rep2 |
     | N2_Day5_rep2     |    N2    | Day5 | rep2 |
     | mir-71_Day0_rep2 |  mir-71  | Day0 | rep2 |
     | mir-71_Day5_rep2 |  mir-71  | Day5 | rep2 |
     | N2_Day0_rep3     |    N2    | Day0 | rep3 |
     | N2_Day5_rep3     |    N2    | Day5 | rep3 |
     | mir-71_Day0_rep3 |  mir-71  | Day0 | rep3 |
     | mir-71_Day5_rep3 |  mir-71  | Day5 | rep3 |

## Install Conda

  - For Linux 64bit, in the terminal:
    ```Bash
      wget https://repo.continuum.io/miniconda/Miniconda2-latest-Linux-x86_64.sh
      sh Miniconda2-latest-Linux-x86_64.sh
    ```
    **During conda's installation: type yes to add conda to the PATH**
    
    For different operating system go to [Conda](https://conda.io/miniconda.html) 
    
## Install NeatSeq-Flow 

  1. Download the **NeatSeq Flow** installer file:
        ```Bash
          wget http://neatseq-flow.readthedocs.io/en/latest/extra/NeatSeq_Flow_conda_env.yaml
        ```
  2. Create the **NeatSeq_Flow** conda environment:
        ```Bash
           conda env create -f NeatSeq_Flow_conda_env.yaml
        ```  
        
## Install DeSeq2 conda environment
  1. Download the **DeSeq2 conda environment** installer file:
        ```Bash
          wget https://raw.githubusercontent.com/bioinfo-core-BGU/NeatSeq-Flow_Workflows/master/DeSeq_Workflow/DeSeq2_module/DeSeq2_env_install.yaml
        ```
  2. Create the **DeSeq2** conda environment:
        ```Bash
          conda env create -f DeSeq2_env_install.yaml
        ```  
   **Note That The Installation Can Take Some Time!!!**
  
        
        
## Install NeatSeq-Flow-GUI
  1. Download the **NeatSeq-Flow-GUI** installer file:
        ```Bash
          wget https://raw.githubusercontent.com/bioinfo-core-BGU/NeatSeq-Flow-GUI/master/NeatSeq_Flow_GUI_installer.yaml
        ```
  2. Create the **NeatSeq_Flow_GUI** conda environment:
        ```Bash
          conda env create -f NeatSeq_Flow_GUI_installer.yaml
        ```  

***
## How To Use The DESeq2 Module 
**The DESeq2 Module Can Be Used in Two Main Ways:**
  1. Using Raw Count Data:
     - Count Matrix of All Samples (**As in this Toturial**)
     - Count Data for Each Sample (Generated By HTSeq-Count or RSEM)
    
  <img src="https://raw.githubusercontent.com/bioinfo-core-BGU/NeatSeq-Flow_Workflows/master/DeSeq_Workflow/doc/Static_Work_Flow.PNG" width="650">
  2. Using Raw Reads Data, as Part of a Work-Flow That Can Generate Count Data (Using HTSeq-Count or RSEM):
    <img src="https://raw.githubusercontent.com/bioinfo-core-BGU/NeatSeq-Flow-GUI/master/doc/Static_GUI.png" width="650">
  
## Start The GUI
  1. Activate the **NeatSeq_Flow_GUI** conda environment:
        ```Bash
          bash
          source activate NeatSeq_Flow_GUI
        ```
  2. Run **NeatSeq_Flow_GUI**:
        ```Bash 
          NeatSeq_Flow_GUI.py
        ```

***
### Add DeSeq2 Step
   <img src="https://raw.githubusercontent.com/bioinfo-core-BGU/NeatSeq-Flow_Workflows/master/DeSeq_Workflow/doc/Add_DeSeq2_Step.gif" width="650">
   
   1. **Add New Step:**   
      In the 'Work-Flow' Tab choose The module template "DeSeq2_Simple(Using_Conda)" and click on the 'Create New Step' button.
   2. **Choose Step base:**
      - Click on the "DeSeq2" step button to see the step options. 
      - Click on the "base" option in the left panel of the GUI, just below the step "DeSeq2" name. 
      - Click on the "Value options:" drop-down menu in the upper left area of the GUI. 
      - Choose the "Merge" step to be the previous step for the DeSeq2 new step.
      - Click on the "Add" button.
      - Click on the "Edit" button.
      
   3. **Edit the "redirects" options of the "DESeq2" step**
   
        - **Below the "redirects" in the left panel of the GUI:**
          - Click on the "--DESIGN" option and edit the "Value" field to : '~ Genotype + Age + Genotype:Age' (**Dont forget the quotation marks**) then click the "Edit" button.
          - Click on the "--LRT" option and edit the "Value" field to : '~ Genotype + Age' (**Dont forget the quotation marks**) then click the "Edit" button.
          - Click on the "--CONTRAST" option and edit the "Value" field to : 'Genotype,N2,mir-71|Age,Day0,Day5' (**Dont forget the quotation marks**) then click the "Edit" button.
          - Click on the "--X_AXIS" option and edit the "Value" field to : 'Age' then click the "Edit" button.
          - Click on the "--GROUP" option and edit the "Value" field to : 'Genotype' then click the "Edit" button.
          - Click on the "--PCA_COLOR" option and edit the "Value" field to : 'Genotype' then click the "Edit" button.
          - Click on the "--PCA_SHAPE" option and edit the "Value" field to : 'Age' then click the "Edit" button.
          
        **All values in this section refer to the fields found in [The Samples Meta-Data File](#the-samples-meta-data-file)**

        **For more information about the values possibilities of "--DESIGN","--LRT" and "--CONTRAST" see the [DESeq2 manual](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)**

        **It is possible to indicate more then one contrast by using "|"**

***
### Possible Options of The "DESeq2" Mudule

It is Possible to Add("New"),Edit or Remove the following Options:  

| General DeSeq2 Module Options 	| Value                                                       	|
|-------------------------------	|-------------------------------------------------------------	|
| use_click                     	| Will use the CLICK clustering program [Shamir et al. 2000](http://www.aaai.org/Papers/ISMB/2000/ISMB00-032.pdf). 	|


| redirects                     	| Value                                                                                                                                                                                   	|
|---------------------------------	|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------	|
| --SAMPLE_DATA_FILE              	| Path to Samples Information File                                                                                                                                                        	|
| --GENE_ID_TYPE                  	| The Gene ID Type i.e 'ENSEMBL'[for Bioconductor] OR 'ensembl_gene_id'/'ensembl_transcript_id' [for ENSEMBL]                                                                             	|
| --Annotation_db                 	| Bioconductor Annotation Data Base Name from https://bioconductor.org/packages/release/BiocViews.html#___OrgDb                                                                           	|
| --Species                       	| Species Name to Retrieve Annotation Data from ENSEMBL                                                                                                                                   	|
| --KEGG_Species                  	| Species Name to Retrieve Annotation Data from KEGG                                                                                                                                      	|
| --KEGG_KAAS                     	| Gene to KO file from KEGG KAAS [first column gene id, second column KO number]                                                                                                          	|
| --Trinotate                     	| Path to a Trinotate annotation file in which the first column is the genes names                                                                                                        	|
| --FILTER_SAMPLES                	| Filter Samples with Low Number of expressed genes OR with Small Library size using 'scater' package                                                                                     	|
| --FILTER_GENES                  	| Filter Low-Abundance Genes using 'scater' package                                                                                                                                       	|
| --NORMALIZATION_TYPE            	| The DeSeq2 Normalization Type To Use [VSD , RLOG] The Default is VSD                                                                                                                    	|
| --BLIND_NORM                    	| Perform Blind Normalization                                                                                                                                                             	|
| --DESIGN                        	| The Main DeSeq2 Design [ ~ Group ]                                                                                                                                                      	|
| --LRT                           	| The LRT DeSeq2 Design                                                                                                                                                                   	|
| --ALPHA                         	| Significant Level Cutoff, The Default is 0.05                                                                                                                                           	|
| --FoldChange                    	| Fold change Cutoff [testing for fold changes greater in absolute value], The Default is 1                                                                                               	|
| --CONTRAST                      	| The DeSeq Contrast Design ["Group,Treatment,Control"] [Not For LTR]. It is possible to define more then one contrast Design ["Group,Treatment1,Control1|Group,Treatment2,Control2|..."] 	|
| --modelMatrixType               	| How the DeSeq model matrix of the GLM formula is formed [standard or expanded] ,The Default is standard                                                                                 	|
| --GENES_PLOT                    	| Genes Id To Plot count Data [separated by ',']                                                                                                                                          	|
| --X_AXIS                        	| The Filed In the Sample Data To Use as X Axis                                                                                                                                           	|
| --GROUP                         	| The Filed In the Sample Data To Group By [can be two fields separated by ',']                                                                                                           	|
| --SPLIT_BY                      	| The Filed In the Sample Data To Split the Analysis By.                                                                                                                                  	|
| --FUNcluster                    	| A clustering function including [kmeans,pam,clara,fanny,hclust,agnes,diana,click]. The default is hclust. If the 'use_click' option is used the '--FUNcluster' option is set to 'click' 	|
| --hc_metric                     	| Hierarchical clustering metric to be used for calculating dissimilarities between observations. The default is pearson                                                                  	|
| --hc_method                     	| Hierarchical clustering agglomeration method to be used. The default is ward.D2                                                                                                         	|
| --k.max                         	| The maximum number of clusters to consider, must be at least two. The default is 20                                                                                                     	|
| --nboot                         	| Number of Monte Carlo (bootstrap) samples for determining the number of clusters [Not For Mclust]. The default is 10                                                                    	|
| --stand                         	| The Data will be Standardized Before Clustering.                                                                                                                                        	|
| --Mclust                        	| Use Mclust for determining the number of clusters.                                                                                                                                      	|
| --CLICK_HOMOGENEITY             	| The HOMOGENEITY [0-1] of clusters using CLICK program (Shamir et al. 2000). The default is 0.5                                                                                          	|
| --PCA_COLOR                     	| The Filed In the Sample Data To Determine Color In The PCA Plot                                                                                                                         	|
| --PCA_SHAPE                     	| The Filed In the Sample Data To Determine Shape In The PCA Plot                                                                                                                         	|
| --PCA_SIZE                      	| The Filed In the Sample Data To Determine Size In The PCA Plot. The default is Library Size                                                                                             	|
| --Enriched_terms_overlap        	| Test for genes overlap in enriched terms                                                                                                                                                	|
| --USE_INPUT_GENES_AS_BACKGROUND 	| Use The input Genes as the Background for Enrichment Analysis                                                                                                                           	|

***
### Edit The Vars Tab:
   <img src="https://raw.githubusercontent.com/bioinfo-core-BGU/NeatSeq-Flow_Workflows/master/DeSeq_Workflow/doc/Edit_Vars.gif" width="650">
    
 **In the 'Vars' Tab:**

   1. Click on the small arrow next to the "DeSeq2" in the lower panel.
   2. Click on "conda env" and edit the "Value" field to 'DeSeq2' and then click on the "Edit" button.
   3. Click on "SAMPLE_DATA" 
      - Click on the "Browse" button.
      - Select the "Samples_MetaData" file.
      - Click on the "OK" button.
      - Click on the "Edit" button.
   4. Click on "Species" and edit the "Value" field to 'Caenorhabditis elegans genes' (**Dont forget the quotation marks**) and then click on the "Edit" button.
   
   **In the "Species" option it is possible to use any of the species names available at [BioMart - Ensembl](https://www.ensembl.org/biomart/martview)***
      
  **Go back to the 'Work-Flow' Tab:**
 
   1. Click on the "Save WorkFlow" button.
   2. Create new file entitled "GSE72232_RNA-seq_DeSeq2_Workflow.yaml".
   3. Click on the "OK" button.

        
***
### Create a Samples File:
   <img src="https://raw.githubusercontent.com/bioinfo-core-BGU/NeatSeq-Flow_Workflows/master/DeSeq_Workflow/doc/Create_Samples_File.gif" width="650">
    
 **In the 'Samples' Tab:**
        
   1. Edit the "Project Title" field in the leftmost corner of the tab to 'GSE72232'.
   2. Click on the "Add Project File" button.
   3. Select the "GSE72232_RNA-seq_combined_counts.txt.gz" file. 
   4. Click on the "Open" button.
   5. Select from the "File Type" drop-down menu the "results" file type.
   6. Click on the "Save Sample File" button.
   7. Create new file entitled "GSE72232_Samples_File"
   8. Click on the "Save" button.
   
 **If you have count Data for each sample separately (Not for this tutorial!! )**
   
   1. Edit the "Project Title" field in the leftmost corner of the tab.
   2. Click on the "Add Sample File" button.
   3. Select all your count data files. 
   4. Click on the "OK" button.
   5. For each sample's count file, edit the file name to match the sample name in the meta-data file.
   6. For each sample's count file, select from the "File Type" drop-down menu:
      - "HTSeq.counts" for HTSeq counts data.
      - "genes.counts" for RSEM counts data.
   7. Click on the "Save Sample File" button.
   8. Create new Samples File.
   9. Click on the "Save" button.
   
   
***
### Run Work-Flow:
   <img src="https://raw.githubusercontent.com/bioinfo-core-BGU/NeatSeq-Flow_Workflows/master/DeSeq_Workflow/doc/Run_DeSeq2.gif" width="850">
    
 **In the 'Run' Tab:**
        
   1. Click on the "Search" button next to the "Conda environment to use" field and wait until at least 2 conda environments are found.
      - If an error occurs try again.
   2. Select from the "Conda environment to use" drop-down menu the "NeatSeq_Flow" environment.
      - If you have previously installed the NeatSeq_Flow Tutorial environment it is also possible to choose it.
   3. Click on the "Browse" button next to the "Parameter File" field.
      - Then select the "GSE72232_RNA-seq_DeSeq2_Workflow.yaml" file and click on the "OPEN" button.
   4. Click on the "Browse" button next to the "Sample File" field.
      - Then select the "GSE72232_Samples_File" file and click on the "OPEN" button.
   5. It is possible to select the output directory by:
      - Clicking on the "Browse" button next to the "Project Directory" field.
      - Then select the your desired output directory and click on the "OPEN" button.
      - Leaving this field empty will set the output directory to the current working directory.
   6. Click on the "Generate scripts" button.
      - If an error occurs try again.
      - If the error re-occurs see the terminal for more info.
   7. Click on the "Run scripts" button.
   8. Click on the "Run Monitor" button.

## Output Data:
   <img src="https://raw.githubusercontent.com/bioinfo-core-BGU/NeatSeq-Flow_Workflows/master/DeSeq_Workflow/doc/Output_Data.PNG" width="850">(https://neatseq-flow.readthedocs.io/en/latest/03.output.html)
 
 *[Click here for more information about the NeatSeq-Flow platform output directory structure](https://neatseq-flow.readthedocs.io/en/latest/03.output.html)**

# Contact
Please contact Liron Levin at: [levinl@post.bgu.ac.il](mailto:levinl@post.bgu.ac.il)
